uqcsbot/api.py

```diff
@@ -24,3 +26,83 @@ def get_members(self) -> List[str]:
             members += page["members"]
         return members
 
+
+class APIMethodProxy(object):
+    """
+    Helper class used to implement APIWrapper
+    """
+    def __init__(self, client: SlackClient, method: str, is_async: bool = False):
+        self._client = client
+        self._method = method
+        self._async = is_async
+
+    def __call__(self, timeout=None, **kwargs) -> Union[dict, Awaitable[dict]]:
+        """
+        Perform the relevant API request. Equivalent to SlackClient.api_call
+        except the `method` argument is filled in.
+
+        If the APIMethodProxy was constructed with `is_async=True` runs the
+        request asynchronously via:
+            asyncio.get_event_loop().run_in_executor(None, req)
+        """
+        fn = partial(
+            self._client.api_call,
+            self._method,
+            timeout=self.timeout,
+            **kwargs
+        )
+        if self._async:
+            loop = asyncio.get_event_loop()
+            return loop.run_in_executor(None, fn)
+        else:
+            return fn()
+
+    def __getattr__(self, item) -> 'APIMethodProxy':
+        """
+        Gets another APIMethodProxy with the same configuration as the current
+        one, except the attribute that you tried to get is appended to the
+        method of the source APIMethodProxy, with a dot separating them.
+
+        For example,
+            > APIMethodProxy("chat").postMessage
+        is equivalent to
+            > APIMethodProxy("chat.postMessage")
+        """
+        return APIMethodProxy(
+            client=self._client,
+            method=f'{self._method}.{item}',
+            is_async=self._async,
+        )
+
+
+class APIWrapper(object):
+    """
+    Wraps the Slack API client to make it possible to use dotted methods. Can
+    perform API requests both synchronously and asynchronously.
+
+    Example usage:
+        > api = APIWrapper(client)
+        > api.chat.postMessage(channel="general", text="message")
+
+        > async_api = APIWrapper(client, True)
+        > await async_api.chat.postMessage(channel="general", text="message")
+
+        > async_api = api(is_async=True)
+        > await async_api.chat.postMessage(channel="general", text="message")
+    """
+    def __init__(self, client: SlackClient, is_async: bool = False):
+        self._client = client
+        self._async = is_async
+
+    def __getattr__(self, item) -> APIMethodProxy:
+        return APIMethodProxy(
+            client=self._client,
+            method=item,
+            is_async=self._async
+        )
+
+    def __call__(self, **kwargs) -> 'APIWrapper':
+        """
+        On call, reconfigure the API Wrapper
+        """
+        return type(self)(self._client, **kwargs)
```

Comment:
What is this doing?

Reply:
This is what enables the following example:
```python
api = APIWrapper(client)
async_api = api(is_async=True)
await async_api.chat.postMessage(channel="general", text="message")
```

It constructs a new object of the same type, passing in the same SlackClient, but otherwise requiring you to pass in any other config. The current behaviour is slightly different to what I intended so I'll update both the code & docs.